### PR TITLE
[FEATURE] Autocompletion

### PR DESCRIPTION
**Pull request by lazybensch**
check out the [RFC](https://github.com/lazybensch/rfcs/blob/add_shell_completion/active/0000-command-line-completion.md) to learn more about this feature.

things left to be done:
- [x] make CI green
- [x] apply code review comments
- [ ] add `fish` support
- [x] decide on cache invalidation technique
- [ ] re-source shell config file

to entertain you until then:

### CODE REVIEW

File: bin/ember [hunk 1]
```diff
@@ -4,6 +4,9 @@
 // Provide a title to the process in `ps`
 process.title = 'ember';
 
+var autocomplete = require('../lib/utilities/cli-command-completion');
```
Comment (stefanpenner) [inline bin/ember]: can this instead be run in cli/index.js ?

changes in `bin/ember` come at a big cost, as `bin/ember` is meant to work across all versions of ember-cli. In essence there is a moratorium on changes to it.

**Comment by stefanpenner** [comment]
awesome

File: bin/generate-cli-commands.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        completionPath = path.join(programFolder, 'completion.sh');
+        if (!fs.existsSync(programFolder)) {
+          fs.mkdirSync(programFolder);
+        }
+        fs.writeFileSync(completionPath, this.generateCompletionCode());
+        fs.appendFileSync(initFile, template('source ' + completionPath));
+        break;
+      case 'zsh':
+        fs.appendFileSync(initFile, template('. <(' + this.program + ' --completion)'));
+    }
```
Comment (stefanpenner) [inline bin/generate-cli-commands.js]: as a `fish` user I would love to have `autocomplete` support for it as well. If you don't have time, I can investigate this myself at a later date.

**Comment by stavarotti** [comment]
:clap:  This looks awesome.

**Comment by stefanpenner** [comment]
@kellyselden @lazybensch whats the word on this.

**Comment by lazybensch** [comment]
@stefanpenner i am still trying to get the specs green. (it may not seem like it though, since i am always amending to my last commit)

Feature works fine on my machine and all but I really have a hard time getting the CI happy. On travis, now there only are a couple of tests red on `NODE=0.10` for whatever reason ^^ `0.12` and `io.js` is green however.

windows is another story. I don't have a windows machine to test on so i am currently using appveyor to test my changes. thats a frustratingly slow process :(

**Comment by lazybensch** [comment]
currently my plan is to get CI green on my own, and then grab someone that can take the time to go through my code with me, double checking that i haven't broken any rules. Since it's a lot of stuff i wrote considering how little i know about ember-cli yet ^^

**Comment by lazybensch** [comment]
the few red tests on node 0.10 seem to be related to https://github.com/npm/fstream/issues/46 and break on more PRs

**Comment by stefanpenner** [comment]
> the few red tests on node 0.10 seem to be related to npm/fstream#46 and break on more PRs
>  Hide all checks

ya i plan to look at that this morning (after breakfast) as it is causing us mega grief.

**Comment by knownasilya** [comment]
What's the status on this?

**Comment by lazybensch** [comment]
@knownasilya i still plan to finish this PR, its working locally - besides some failing tests - so there is not that much to do anymore. currently i'm on holidays but am looking forward to finalise this afterwards.

File: lib/models/addon-discovery.js [hunk 3]
```diff
@@ -102,7 +102,7 @@ AddonDiscovery.prototype.discoverFromProjectItself = function(project) {
 AddonDiscovery.prototype.discoverFromDependencies = function(root, nodeModulesPath, pkg, excludeDevDeps) {
   var discovery = this;
   var addons = Object.keys(this.dependencies(pkg, excludeDevDeps)).map(function(name) {
-    if (name !== 'ember-cli') {
+    if (['ember-cli', 'github'].indexOf(name) === -1) {
```
Comment (stefanpenner) [inline lib/models/addon-discovery.js]: can you share the reasoning for this, it isn't obvious to me.

File: lib/models/cli-command-completer.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  @extends CoreObject
+  @constructor
+*/
+
+function CliCommandCompleter() {
+
+  try {
+    this.completion = omelette('ember');
+    this.template = this.loadTemplate('/node_modules/ember-cli/cli-commands.json');
+  } catch(e) {}
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: do we expect this ever to fail?

File: lib/models/cli-command-completer.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+  @class CliCommandCompleter
+  @extends CoreObject
+  @constructor
+*/
+
+function CliCommandCompleter() {
+
+  try {
+    this.completion = omelette('ember');
+    this.template = this.loadTemplate('/node_modules/ember-cli/cli-commands.json');
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: This path should be relative the current dirname, not the process.

File: lib/models/cli-command-completer.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  content. If no cache file exists it will fallback to generate the JSON
+  on demand.
+
+  @private
+  @method loadTemplate
+ */
+
+CliCommandCompleter.prototype.loadTemplate = function(relPathToCache) {
+
+  var pathToCache = process.cwd() + relPathToCache;
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: we must load stuff relative to the dirname, not process. As ember-cli can be run deeply nested..

File: lib/models/cli-command-completer.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+CliCommandCompleter.prototype.handleInput = function(fragment, word, line) {
+
+  var intent = {
+    arguments: this.getArgs(line),
+    isPremature: this.isPremature(line),
+    seeksOptions: this.seeksOptions(line)
+  };
+
+  try {
+    this.completion.reply(this.getResult(intent));
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: When does this error? I suspect we don't actually want to swallow all exceptions this code-path could throw

File: lib/models/cli-command-completer.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+  @private
+  @method getCommand
+  @param  {Object}  intent Represents intent of user
+  @return {Command}        Matching command. `null` otherwise.
+ */
+
+CliCommandCompleter.prototype.getCommand = function(intent) {
+  var command = intent.arguments.reduce(function(parentCommand, term) {
+
+    var command = this.findCommand(parentCommand.commands, term);
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: funky whitespace

File: lib/models/cli-command-completer.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+    if (!command) {
+      return intent.isPremature ? parentCommand : {};
+    }
+
+    // remember options of parent command
+    command.options = command.options.concat(parentCommand.options || []);
+    return command;
+
+  }.bind(this), this.template );
```
Comment (stefanpenner) [inline lib/models/cli-command-completer.js]: trailing whitespace after `te`

File: lib/models/cli-command-generator.js [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+function CliCommandGenerator(project) {
+  try {
+    this.project = this.prepareProject(project);
+    this.cacheDir = '/node_modules/ember-cli/';
+    this.cacheFile = 'cli-commands.json';
+    this.blueprints = loadBlueprints(this.project);
+    this.commands = loadCommands(this.project);
+    this.root = loadRoot();
+  } catch(e) {}
```
Comment (stefanpenner) [inline lib/models/cli-command-generator.js]: When does this fail, and do we really want to ignore all failures?

File: lib/models/cli-command-generator.js [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+  try {
+    var pathToCache = this.project.root + this.cacheDir + this.cacheFile;
+    var json = this.generateJSON(this.root);
+
+    this.openOrCreateCache(this.project.root + this.cacheDir, this.cacheFile);
+    this.setupShellCompletion();
+
+    var result = JSON.stringify(json, null, ' ');
+    fs.writeFileSync(pathToCache, result);
+  } catch(e) {}
```
Comment (stefanpenner) [inline lib/models/cli-command-generator.js]: ^

**Comment by stefanpenner** [comment]
@lazybensch I'm very much +1 on getting this in. Is their anything I can do to help you make some progress?

**Comment by stefanpenner** [comment]
One thing i noticed, there is alot of very specific to auto-complete code here, should we extract it to a separate module? I can provide a repo on the ember-cli org if desired.

**Comment by lazybensch** [comment]
hey @stefanpenner, i'm glad to hear that :)
as you see i left a lot of debugging code in the PR from when I was trying to find the cause of a failing test (i think it turned out later that master was broken though) anyway i kinda lost my ambition afterwards :(

How about the following: I will get back into the code, clean everything up and rebase. Once I have familiarised myself with this PR again (it's been stale a few month after all) i will figure out what my blockers were back then and tell you if and how you can help me.

I expect to be able to come back to you with some action points *latest* saturday.

File: lib/models/cli-command-completer.js (hunk 6 above)
Comment (lazybensch) [inline lib/models/cli-command-completer.js]: oh crap, perfect! if i remember correctly this was causing me quite some grief with the test suite. thx!

**Comment by stefanpenner** [comment]
> How about the following: I will get back into the code, clean everything up and rebase. Once I have familiarised myself with this PR again (it's been stale a few month after all) i will figure out what my blockers were back then and tell you if and how you can help me.
> 
> I expect to be able to come back to you with some action points latest saturday.

sounds great, keep me posted!

**Comment by Turbo87** [comment]
@lazybensch any news on this? can we help move this forward in some way?

**Comment by Turbo87** [comment]
FYI I've rebased this branch and fixed a few issues: https://github.com/Turbo87/ember-cli/tree/autocompletion

File: lib/models/cli-command-completer.js (hunk 8 above)
Comment (lazybensch) [inline lib/models/cli-command-completer.js]: are you referring to the indent? line 150 starts a function body.

**Comment by lazybensch** [comment]
@stefanpenner @Turbo87 lets make this feature happen

what i've done the last two days:
- branch is rebased
- code review comments addressed (especially moving code out of `bin/ember`)
- CI appears green (hard to tell though for appVeyor)
- feature currently works for bash and zsh (tested only with macOS on 2 machines)

You can consider this PR done, it works and everything **but** there are some decisions i made and would like to hear your opinion about:
  1. **shell resourcing:** To get this feature working you need to execute the ember command once and it will append one line to your shell config that enables autocompletion for you. Afterwards you need to re-source your shell (or re-open it) to ultimately make use of the shell completion. The need to re-source the shell is something i can not work around, it would prefer the user could just use this feature immediately but i'm quite sure this is not possible. Correct me if i'm wrong.

  2. **command caching:** This feature does not only autocomplete standard ember-cli commands but also all commands that are introduced by addons. Parsing all those addons for potential commands is somewhat slow (faster then a second but to slow for tab-completion) so i am generating a cache file that stores all the commands and their options and what not. This makes the tab-completion satisfyingly fast.

  3. **cache invalidation:** Point 2 raises the need for cache invalidation. For example when a user adds a new addon with new commands, the autocompletion should reflect this change. I decided to invalidate the cache every time you **execute** an ember command. Since i only need to rewrite the cache file without reading it afterwards i can do this asynchronously so executing the ember command suffers no performance penalty. This however means that when a collaborator adds an addon and you do `git pull; npm i;` and you immediately try the tab completion, the new commands from this addon wont be shown. Only after you execute some command (like `ember version` or `ember serve`) and then try the autocompletion again, the cache file is updated and you'll also see the newly added commands.

  4. **separate module:** as @stefanpenner mentioned earlier, it might makes sense to extract the logic into a dedicated module. Im technically in favour of this **but** since this PR dangled in lazyness-limbo for almost 6 month i would prefer getting it merged in ASAP. I feel confident that - assuming we will merge this soon - I can extract its logic into an external repo as a "second iteration".

I'm so sorry nothing happened here for so long, especially now that i know there wasn't that much left to be done ^^

**Comment by lazybensch** [comment]
also, if you need to discuss anything in detail or have questions - do not hesitate to reach me on slack.

**Comment by lazybensch** [comment]
things left to do:
- [ ] node 0.10 build is still failing
- [ ] currently no windows support

 windows support is a bit tricky since i've no windows machine but i am making some progress there anyway.

**Comment by kellyselden** [comment]
@lazybensch You can run the VM at https://github.com/ember-cli/ember-cli/tree/master/dev/windows if you need a machine.

**Comment by lazybensch** [comment]
perfect, thx

**Comment by Turbo87** [comment]
or use https://github.com/xdissent/ievms ;)

**Comment by knownasilya** [comment]
I'd also add a command to manually invalidate the cache, but that's just me.

**Comment by lazybensch** [comment]
@knownasilya even though **every** command you execute is invalidating the cache anyways?

**Comment by Turbo87** [comment]
@lazybensch can you explain why you introduced the CliCommand class? does that offer anything that the Command class isn't offering?

also why did you introduce another JSON manifest format thingy instead of reusing the output of `ember help --json`? I could imagine that if that command was cached the same way that you proposed it would have other benefits (e.g. for IDE integration).

**Comment by knownasilya** [comment]
> Ilya Radchenko even though every command you execute is invalidating the cache anyways?

This way you could create a postinstall that invalidates the cache, and you never have to deal with the install issue.

**Comment by lazybensch** [comment]
@knownasilya wouldn't a postinstall only invalidate the cache on the contributor machine that is adding an addon via `ember install addon-with-commands`? when a collaborator does `git pull; npm i` he will have an invalid cache, right?

**Comment by knownasilya** [comment]
No, because `postinstall` runs after any npm install in that project, so it will but the consumer has to create the postinstall on their end (in their ember app package.json).

**Comment by lazybensch** [comment]
@knownasilya well thx for the suggestion then. I'll implement it the way you described.

@Turbo87 a CliCommand is, despite the similar name, a semantically different thing (for example do also blueprints need a CliCommand representation). This Class describes the actual words that are displayed on the command line and what words should follow etc.

The `help --json` was not implemented back when i first started this PR. It's current output has some format properties that do not play along with command line completion and of course its generation takes to long. I can, however make the `help --json` command read the cache file instead (i think this is what you proposed in the second half of your comment).

I have not touched this so far because, well... you see this feature is way more involved as i first anticipated and i would have prefered to "not care" about this for now, get a mergeable MVP, and DRY-ing up the json generation in a later PR but just now as i wrote this thought down i already realize that this is not a good idea :D 

also i've red *IDE integration* a couple of times in context of `help --json`, what is that about? can you give me an example?

**Comment by knownasilya** [comment]
It might even make sense to add the postinstall to new projects via the app blueprint in this PR.

**Comment by Turbo87** [comment]
> also i've red IDE integration a couple of times in context of help --json, what is that about? can you give me an example?

take https://github.com/Turbo87/intellij-emberjs for example. it allows you to run `ember generate` from the UI, but it currently has a hardcoded list of available blueprints. if there was something like a `<project>/node_modules/ember-cli/.help.json` the plugin could replace the hardcoded list with the blueprints that are actually supported in the project similar to how the autocompletion would show a list of supported blueprints.

> The help --json was not implemented back when i first started this PR.

good point. but it is implemented now, so IMHO we should try to reuse what already exists.

> It's current output has some format properties that do not play along with command line completion

can you give examples of what isn't playing along?

> and of course its generation takes to long

thus my suggestion to cache the output similar to what you're suggesting in this PR

> I can, however make the help --json command read the cache file instead (i think this is what you proposed in the second half of your comment).

I actually would like to propose the opposite: cache the output of `help --json` and use that for the autocompletion instead of introducing another JSON commands cache file format

**Comment by lazybensch** [comment]
well, here are a few arguments why i would prefere using the cli-command-generator:

- When the `help --json` code was written, it kind of made sense to put it inside `lib/commands/help.js`  since it was directly related to this command. Now that it would generate a cache file that is both used  for autocompletion **and** help output (especially now that i know about IDE integration) it should be  decoupled from the help command and move into a seperate module. Decoupling the currently existing json generation from the `help.js` would be quite some effort. Especially  considering that i already provide this with `lib/models/cli-command-generator.js` which sole responsible is to generate that json.

- As I mentioned there are a few inconsistencies in the way json needs to be formatted for either the help
  output or the cli-completion. For example shell completion does not discriminate between blueprints and commands  or native commands and addon commands. (I mean of course this can be added as meta information but  should not effect the structure of the command tree). I believe IDE integration would also profit from the  structure used for autocompletion.

- This is quite a weak argument but still: cli-command-generator is fully documented

- The JSON generated by `help --json` does not mention that you can do `ember help generate model`. However this  information needs to be represented when you want a capable autocompletion. Again, i believe IDE integration  would profit from this aswell.

- To future proof this feature it is important to have a good API for commands and blueprints (build-in
  ones and those of addon authors) the `cliCommands` function that you can override in your command **or  blueprint** passes you, in case of autocompletion, what cliCommands are already typed (example below). This might  not seem like much at first glance but it means that addon authors will have partial control over how there  commands should be completed. For example the `help` command overrides `cliCommands` to see if it was already  mentioned like so:
  ```javascript
    cliCommands: function(generator, parentCommands) {
      if (parentCommands.indexOf('help') > -1) {
        return [];
      } else {
        return generator.commands;
      }
  ```
  The effect is that when you do `ember <tab>` your shell will list `help`, among others, as a suggestion, and if you  do `ember help <tab>` it will again, suggest `help` (since `ember help help` is a valid command) but once you do `ember help help` or `ember generate help` you will not get `help as a suggestion anymore`.

I believe the last point is bigger then it seems. Because addon authors are able to help users with suggestions.For example imagine ember-deploy would be a monolithic addon (instead of using plugins) that would let you deploy to any destination. The addon introduces a `ember deploy` command, you invoke it by passing a destination. The addon author could now override the `cliCommands` function of his `deploy` method to suggest ['s3', 'redis', 'heroku']. A user could then do `ember deploy <tab>` and gets all suggestions, or simply do `ember deploy he<tab>` to safe time. Another usecase is that you can easily add deeply nested commands to your addon if you want to.

**Comment by Turbo87** [comment]
> Decoupling the currently existing json generation from the help.js would be quite some effort.

I totally agree on that, but I still think it's worth the effort. The alternative of having essentially two JSON outputs (the `ember help --json` one and yours) feels kinda dirty and confusing to me.

> For example shell completion does not discriminate between blueprints and commands or native commands and addon commands.

I think it should be reasonably easy to convert the current `help --json` output to the command tree that you're proposing.

> To future proof this feature it is important to have a good API for commands and blueprints (build-in ones and those of addon authors) the cliCommands function that you can override in your command or blueprint

correct me if I'm wrong, but I think we could still do the same thing if we used the other JSON output. we would need to figure out if the last "cli command" was a command or blueprint, look it up and then add the `cliCommands()` result to the list of completions.

**Comment by lazybensch** [comment]
I hope i am not wearing you out with this discussion but there is one point i think i haven't made clear:
I don't want to have two JSON generations, i too definitely only want to keep one. So naturally there are two options:

**option A**
- remove JSON generation in `command/help.js`
- replace it with with a function that maps *my* cached json into the help format
- profit

**option B**
- remove JSON generation in `cli-command-generator.js``
- make the completer map the *help --json* cached structure into autocomplete format
- refactor the code in `commands/help.js` into its own module
- make it possible to control subcommands based on already typed commands
- make the help command use this api to reproduce the behavior i explained earlier
- document that new module
- profit

Not only is option B a lot more effort on my side to have the same result; I argue that option A and option B have the **exact same result**. Actually if i squash my commits i bet you could not tell which option i chose as following option B would result in a decoupled documented module that supports the features i mentioned. But that is **exactly** what i already have implemented in `cli-command-generator.js`.

**Comment by Turbo87** [comment]
I guess it boils down to the question which of the formats is more generic and can be used by more than just the autocompletion code. I would personally answer this question with the `ember help --json` output because I think it better represents the internal structure, explicitly mentions the addons, etc.

**Comment by lazybensch** [comment]
I still feel adding those properties to the autocompletion could would be less effort then adding the missing properties to the help code. also i am not quite sure if i can produce the structure needed for autocompletion out of the help-json but think the other way around is not that hard (as i said, i would add addon and blueprint information as meta data to all the cliCommands in my hash

**Comment by nathanhammond** [comment]
Added to tomorrow's ember-cli core team meeting agenda.

**Comment by Turbo87** [comment]
@lazybensch can you change this PR to use the `json-generator` utility for now until #5635 is merged? I'd rather have slow completion than now no completion at all :wink:

**Comment by sobi3ch** [comment]
So what's happening here, is this issue going anywhere?

**Comment by Turbo87** [comment]
closing this for now due to missing activity. would love it if someone would pick it up again though.

**Pull request closed without merging**